service/worker/batcher/batcher.go

```diff
@@ -0,0 +1,174 @@
+	// the batcher sub-system
+	BootstrapParams struct {
+		// Config contains the configuration for scanner
+		Config Config
+		// ServiceClient is an instance of cadence service client
+		ServiceClient workflowserviceclient.Interface
+		// MetricsClient is an instance of metrics object for emitting stats
+		MetricsClient metrics.Client
+		Logger        log.Logger
+		// TallyScope is an instance of tally metrics scope
+		TallyScope tally.Scope
+	}
+
+	// batcherContext is the context object that get's
+	// passed around within the scanner workflows / activities
+	batcherContext struct {
+		cfg           Config
+		svcClient     workflowserviceclient.Interface
+		metricsClient metrics.Client
+		tallyScope    tally.Scope
+		logger        log.Logger
+	}
+
+	// Batcher is the background sub-system that execute workflow for batch operations
+	Batcher struct {
+		context batcherContext
+	}
+)
+
+// New returns a new instance of batcher daemon Batcher
+func New(params *BootstrapParams) *Batcher {
+	cfg := params.Config
+	return &Batcher{
+		context: batcherContext{
+			cfg:           cfg,
+			svcClient:     params.ServiceClient,
+			metricsClient: params.MetricsClient,
+			tallyScope:    params.TallyScope,
+			logger:        params.Logger.WithTags(tag.ComponentBatcher),
+		},
+	}
+}
+
+// Start starts the scanner
+func (s *Batcher) Start() error {
+	//retry until making sure global system domain is there
+	err := s.createGlobalSystemDomainIfNotExistsWithRetry()
+	if err != nil {
+		return err
+	}
+
+	// start worker for batch operation workflows
+	workerOpts := worker.Options{
+		MetricsScope:              s.context.tallyScope,
+		BackgroundActivityContext: context.WithValue(context.Background(), batcherContextKey, s.context),
+	}
+	worker := worker.New(s.context.svcClient, common.SystemDomainName, batcherTaskListName, workerOpts)
+	return worker.Start()
+}
+
+func (s *Batcher) createGlobalSystemDomainIfNotExistsWithRetry() error {
+	policy := backoff.NewExponentialRetryPolicy(backOffInitialInterval)
+	policy.SetMaximumInterval(backOffMaxInterval)
+	policy.SetExpirationInterval(maxStartupTime)
+	return backoff.Retry(func() error {
+		return s.createGlobalSystemDomainIfNotExists()
+	}, policy, func(err error) bool {
+		return true
+	})
+}
+
+func (s *Batcher) createGlobalSystemDomainIfNotExists() error {
+	ctx, cancel := context.WithTimeout(context.Background(), rpcTimeout)
+	_, err := s.context.svcClient.DescribeDomain(ctx, &shared.DescribeDomainRequest{
+		Name: common.StringPtr(common.SystemGlobalDomainName),
+	})
+	cancel()
+
+	if err == nil {
+		s.context.logger.Info("Global system domain already exists", tag.WorkflowDomainName(common.SystemGlobalDomainName))
+		return nil
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), rpcTimeout)
+	err = s.context.svcClient.RegisterDomain(ctx, s.getDomainCreationRequest())
+	cancel()
+	if err != nil {
+		s.context.logger.Error("Error creating global system domain", tag.Error(err))
```

Comment:
maybe should check if domain not active error is returned here

depends on how you want to deal with global domain creation, there can be several things
1. if seeing domain not active here, loop + backoff + retry on describe domain, waiting on worker to create the domain on active side
2. crash.

Reply:
Sure

Reply:
the error we are getting is:
```
errNotMasterCluster                = &gen.BadRequestError{Message: "Cluster is not master cluster, cannot do domain registration or domain update."}
```
do you want to check on badRequest and the message inside?